### PR TITLE
[sml] add KNN in jax

### PR DESCRIPTION
**Pull request by hacker-jerry**
完成了初步的knn算法，并且通过了sispm的unittest。
但是在运行ppd的时候，出现了报错。
```
Traceback (most recent call last):
  File "simple_knn_emul.py", line 94, in <module>
    emul_SimplePCA(emulation.Mode.MULTIPROCESS)
  File "simple_knn_emul.py", line 68, in emul_SimplePCA
    result = emulator.run(proc)(X_train, y_train_new, X_test, n_classes, n_neighbors=3)(static_argnums=(3,4))
  File "/Users/matr1x/Codes/GitHub/spu/sml/knn/../../sml/utils/emulation.py", line 180, in wrapper
    res = ppd.device("SPU")(func)(*args, **kwargs)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/distributed.py", line 655, in __call__
    executable, args_flat, out_tree = self._compile_jax_func(
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/distributed.py", line 724, in _compile_jax_func
    executable, output = spu_fe.compile(
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/frontend.py", line 112, in compile
    ir_text, output = _jax_compilation(fn, static_argnums, args, kwargs)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/cachetools/__init__.py", line 737, in wrapper
    v = func(*args, **kwargs)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/frontend.py", line 41, in _jax_compilation
    cfn, output = jax.xla_computation(
  File "simple_knn_emul.py", line 45, in proc
    return knn_model.transform(X_test, n_neighbors=n_neighbors, weights=weights, n_classes=n_classes)
  File "/Users/matr1x/Codes/GitHub/spu/sml/knn/../../sml/knn/simple_knn.py", line 72, in transform
    prediction = self._predict_single_sample(x, n_neighbors, weights, n_classes)
  File "/Users/matr1x/Codes/GitHub/spu/sml/knn/../../sml/knn/simple_knn.py", line 47, in _predict_single_sample
    k_indices = sorted_indices[:n_neighbors]
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/array_methods.py", line 791, in op
    return getattr(self.aval, f"_{name}")(self, *args)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4143, in _rewriting_take
    return _gather(arr, treedef, static_idx, dynamic_idx, indices_are_sorted,
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4152, in _gather
    indexer = _index_to_gather(shape(arr), idx)  # shared with _scatter_update
  File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4404, in _index_to_gather
    raise IndexError(msg)
IndexError: Array slice indices must have static start/stop/step to be used with NumPy indexing syntax. Found slice(None, Traced<ShapedArray(int32[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>, None). To index a statically sized array at a dynamic position, try lax.dynamic_slice/dynamic_update_slice (JAX does not support dynamically sized arrays within JIT compiled functions).
```
我认为这是不应该发生的，因为在sispm进行 jit 编译的时候并没有出错，麻烦老师查看一下具体的错误原因，谢谢！

### CODE REVIEW

**Comment by deadlywing** [comment]
> 完成了初步的knn算法，并且通过了sispm的unittest。 但是在运行ppd的时候，出现了报错。
> 
> ```
> Traceback (most recent call last):
>   File "simple_knn_emul.py", line 94, in <module>
>     emul_SimplePCA(emulation.Mode.MULTIPROCESS)
>   File "simple_knn_emul.py", line 68, in emul_SimplePCA
>     result = emulator.run(proc)(X_train, y_train_new, X_test, n_classes, n_neighbors=3)(static_argnums=(3,4))
>   File "/Users/matr1x/Codes/GitHub/spu/sml/knn/../../sml/utils/emulation.py", line 180, in wrapper
>     res = ppd.device("SPU")(func)(*args, **kwargs)
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/distributed.py", line 655, in __call__
>     executable, args_flat, out_tree = self._compile_jax_func(
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/distributed.py", line 724, in _compile_jax_func
>     executable, output = spu_fe.compile(
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/frontend.py", line 112, in compile
>     ir_text, output = _jax_compilation(fn, static_argnums, args, kwargs)
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/cachetools/__init__.py", line 737, in wrapper
>     v = func(*args, **kwargs)
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/frontend.py", line 41, in _jax_compilation
>     cfn, output = jax.xla_computation(
>   File "simple_knn_emul.py", line 45, in proc
>     return knn_model.transform(X_test, n_neighbors=n_neighbors, weights=weights, n_classes=n_classes)
>   File "/Users/matr1x/Codes/GitHub/spu/sml/knn/../../sml/knn/simple_knn.py", line 72, in transform
>     prediction = self._predict_single_sample(x, n_neighbors, weights, n_classes)
>   File "/Users/matr1x/Codes/GitHub/spu/sml/knn/../../sml/knn/simple_knn.py", line 47, in _predict_single_sample
>     k_indices = sorted_indices[:n_neighbors]
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/array_methods.py", line 791, in op
>     return getattr(self.aval, f"_{name}")(self, *args)
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4143, in _rewriting_take
>     return _gather(arr, treedef, static_idx, dynamic_idx, indices_are_sorted,
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4152, in _gather
>     indexer = _index_to_gather(shape(arr), idx)  # shared with _scatter_update
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 4404, in _index_to_gather
>     raise IndexError(msg)
> IndexError: Array slice indices must have static start/stop/step to be used with NumPy indexing syntax. Found slice(None, Traced<ShapedArray(int32[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>, None). To index a statically sized array at a dynamic position, try lax.dynamic_slice/dynamic_update_slice (JAX does not support dynamically sized arrays within JIT compiled functions).
> ```
> 
> 我认为这是不应该发生的，因为在sispm进行 jit 编译的时候并没有出错，麻烦老师查看一下具体的错误原因，谢谢！

hello，这个问题是因为emulator不支持`static_argnums`，spu的同学最近应该会fix这个问题。（fix以后再通知您）

当前可以先用partial把那俩参数先固定就可以正常运行啦～

**Comment by deadlywing** [comment]
您好，我们最新的commit里已经在emulator里增加了 `static_argnums` 这个参数，您可以再试一试～

Thanks

**Comment by tpppppub** [comment]
> > 您好，我们最新的commit里已经在emulator里增加了 `static_argnums` 这个参数，您可以再试一试～
> > Thanks
> 
> 您好，测试之后还是有点问题的，具体报错如下：
> 
> ```
> Traceback (most recent call last):
>   File "simple_knn_emul.py", line 94, in <module>
>     emul_SimplePCA(emulation.Mode.MULTIPROCESS)
>   File "simple_knn_emul.py", line 68, in emul_SimplePCA
>     result = emulator.run(proc,static_argnums=(3,4))(X_train, y_train_new, X_test, n_classes, n_neighbors=3)
>   File "/Users/matr1x/Codes/GitHub/spu/sml/knn/../../sml/utils/emulation.py", line 170, in wrapper
>     res = ppd.device("SPU")(func, static_argnums, copts)(*args, **kwargs)
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/distributed.py", line 657, in __call__
>     executable, args_flat, out_tree = self._compile_jax_func(
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/spu/utils/distributed.py", line 710, in _compile_jax_func
>     _, dyn_args = japi_util.argnums_partial_except(
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/api_util.py", line 275, in argnums_partial_except
>     static_argnums = _ensure_inbounds(allow_invalid, len(args), static_argnums)
>   File "/opt/homebrew/Caskroom/miniforge/base/envs/sf/lib/python3.8/site-packages/jax/_src/api_util.py", line 262, in _ensure_inbounds
>     raise ValueError(
> ValueError: Positional argument indices, e.g. for `static_argnums`, must have value greater than or equal to -len(args) and less than len(args), but got value 4 for len(args) == 4.
> ```
> 
> 但是代码是没有问题的，arg应该长度为 5，不知道为什么这里显示len(args)=4,麻烦您再排查一下了，谢谢！

因为你的 n_neighbors 是 kwargs，不是 args

File: sml/pca/simple_pca_emul.py [hunk 1]
```diff
@@ -63,7 +63,8 @@ def proc_reconstruct(X):
         emulator.up()
         # Create a simple dataset
         X = random.normal(random.PRNGKey(0), (15, 100))
-        result = emulator.run(proc)(X)
+        X_ = emulator.seal(X)
+        result = emulator.run(proc)(X_)
```
Comment (deadlywing) [inline sml/pca/simple_pca_emul.py:67]: 这里没问题

但是后面` result = emulator.run(proc_reconstruct)(X)`这里忘记修改了哈～

File: sml/knn/simple_knn.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            weighted_counts = jnp.bincount(k_labels, weights=weights_inner, length=int(n_classes))
+            prediction = jnp.argmax(weighted_counts)
+            return prediction
+
+        return lax.cond(weights == 'uniform',
+                        predict_uniform,
+                        predict_distance,
+                        operand=None)
+
+    def transform(self, X, n_neighbors, weights, n_classes):
```
[the remark this answers is not in the document]
Comment (deadlywing) [reply sml/knn/simple_knn.py:69]: weights，n_neighbors,n_classes这种超参数还是移动到__init__方法里吧？（同理，麻烦comment里增加一下当前超参数的含义）

我们希望我们的api能尽可能和sklearn保持一致哈～

File: sml/knn/simple_knn.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+import jax
+from jax import jit, random
+from functools import partial
+import jax.numpy as jnp
+import numpy as np
+from jax import vmap, lax
+from collections import defaultdict
+
+class KNN:
```
Comment (deadlywing) [inline sml/knn/simple_knn.py:23]: 当前实现我理解是一个Classifer，命名上可以明确一下～

File: sml/knn/simple_knn.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+        return lax.cond(weights == 'uniform',
+                        predict_uniform,
+                        predict_distance,
+                        operand=None)
+
+    def transform(self, X, n_neighbors, weights, n_classes):
+        predictions = []
+        for x in X:
+            prediction = self._predict_single_sample(x, n_neighbors, weights, n_classes)
```
Comment (deadlywing) [inline sml/knn/simple_knn.py:72]: 这里可以考虑利用 vmap 进行并行化而不是python层的for loop～

File: sml/knn/simple_knn.py (hunk 3 above)
Comment (deadlywing) [reply sml/knn/simple_knn.py:23]: 我觉得您可以直接在comment doc里直接说明y需要是形如0,1,2...的label数据，让用户明确了解需要如何预处理

**Pull request closed without merging**